File: frontend/src/lib/instagram.ts
import { useQuery } from '@tanstack/react-query';
import { CanonicalPost } from './connectors';

// Instagram Graph API Types (Read-only without Business account)

export interface InstagramUser {
  id: string;
  username: string;
  name?: string;
  biography?: string;
  profile_picture_url?: string;
  website?: string;
  followers_count: number;
  follows_count: number;
  media_count: number;
  account_type: 'BUSINESS' | 'MEDIA_CREATOR' | 'PERSONAL';
}

export interface InstagramMedia {
  id: string;
  caption?: string;
  media_type: 'IMAGE' | 'VIDEO' | 'CAROUSEL_ALBUM';
  media_url?: string;
  permalink: string;
  thumbnail_url?: string;
  timestamp: string;
  username?: string;
  like_count?: number;
  comments_count?: number;
  children?: InstagramMediaChild[];
}

export interface InstagramMediaChild {
  id: string;
  media_type: 'IMAGE' | 'VIDEO';
  media_url: string;
}

export interface InstagramInsights {
  media_id: string;
  impressions: number;
  reach: number;
  engagement: number;
  saved: number;
  video_views?: number;
}

export interface InstagramComment {
  id: string;
  text: string;
  timestamp: string;
  username: string;
  like_count: number;
}

export interface InstagramStory {
  id: string;
  media_type: 'IMAGE' | 'VIDEO';
  media_url: string;
  timestamp: string;
}

export interface InstagramMediaResponse {
  data: InstagramMedia[];
  next_cursor?: string;
}

export interface InstagramStoriesResponse {
  data: InstagramStory[];
}

// Utility Functions

export function getProfileUrl(username: string): string {
  return `https://instagram.com/${username}`;
}

export function extractHashtags(caption: string): string[] {
  const hashtagRegex = /#[a-zA-Z0-9_]+/g;
  return (caption.match(hashtagRegex) || []).map((h) => h.slice(1));
}

export function instagramToCanonical(media: InstagramMedia, user: InstagramUser): CanonicalPost {
  const mediaItems = media.media_type === 'CAROUSEL_ALBUM' && media.children
    ? media.children.map((child) => ({
        type: child.media_type === 'VIDEO' ? 'video' as const : 'image' as const,
        url: child.media_url,
      }))
    : media.media_url
      ? [{ type: media.media_type === 'VIDEO' ? 'video' as const : 'image' as const, url: media.media_url }]
      : undefined;

  return {
    id: media.id,
    content: media.caption || '',
    created_at: media.timestamp,
    author: {
      id: user.id,
      handle: user.username,
      displayName: user.name || user.username,
      avatar: user.profile_picture_url,
    },
    media: mediaItems,
    metrics: { likes: media.like_count || 0, reposts: 0, replies: media.comments_count || 0, quotes: 0 },
  };
}

// API Client

class InstagramApiClient {
  private async request<T>(endpoint: string, options: RequestInit = {}): Promise<T> {
    const response = await fetch(`/api/v1/instagram${endpoint}`, {
      ...options,
      headers: {
        'Content-Type': 'application/json',
        ...options.headers,
      },
      credentials: 'include',
    });

    const data = await response.json();

    if (!response.ok) {
      throw new Error(data.error || `HTTP ${response.status}`);
    }

    return data.data;
  }

  async getOwnProfile(): Promise<InstagramUser> {
    return this.request('/profile');
  }

  async getProfileByUsername(username: string): Promise<InstagramUser> {
    return this.request(`/profile/${encodeURIComponent(username)}`);
  }

  async getMedia(limit = 25, cursor?: string): Promise<InstagramMediaResponse> {
    const params = new URLSearchParams();
    params.set('limit', String(limit));
    if (cursor) params.set('after', cursor);
    return this.request(`/media?${params.toString()}`);
  }

  async getSingleMedia(mediaId: string): Promise<InstagramMedia> {
    return this.request(`/media/${encodeURIComponent(mediaId)}`);
  }

  async getMediaInsights(mediaId: string, mediaType?: string): Promise<InstagramInsights> {
    const params = new URLSearchParams();
    if (mediaType) params.set('media_type', mediaType);
    const query = params.toString();
    return this.request(`/media/${encodeURIComponent(mediaId)}/insights${query ? `?${query}` : ''}`);
  }

  async getStories(): Promise<InstagramStoriesResponse> {
    return this.request('/stories');
  }

  async getAccountInsights(period = 'day'): Promise<{ period: string; insights: Record<string, number> }> {
    return this.request(`/account/insights?period=${period}`);
  }
}

export const instagramApi = new InstagramApiClient();

// Hooks

export function useInstagramOwnProfile() {
  return useQuery<InstagramUser>({
    queryKey: ['instagram-own-profile'],
    queryFn: () => instagramApi.getOwnProfile(),
  });
}

export function useInstagramProfile(username?: string) {
  return useQuery<InstagramUser>({
    queryKey: ['instagram-profile', username],
    queryFn: () => instagramApi.getProfileByUsername(username!),
    enabled: !!username,
  });
}

export function useInstagramMedia(userId?: string, cursor?: string) {
  return useQuery<InstagramMediaResponse>({
    queryKey: ['instagram-media', userId, cursor],
    queryFn: () => instagramApi.getMedia(25, cursor),
    enabled: !!userId,
  });
}

export function useInstagramSingleMedia(mediaId?: string) {
  return useQuery<InstagramMedia>({
    queryKey: ['instagram-single-media', mediaId],
    queryFn: () => instagramApi.getSingleMedia(mediaId!),
    enabled: !!mediaId,
  });
}

export function useInstagramInsights(mediaId?: string, mediaType?: string) {
  return useQuery<InstagramInsights>({
    queryKey: ['instagram-insights', mediaId, mediaType],
    queryFn: () => instagramApi.getMediaInsights(mediaId!, mediaType),
    enabled: !!mediaId,
  });
}

export function useInstagramStories(userId?: string) {
  return useQuery<InstagramStoriesResponse>({
    queryKey: ['instagram-stories', userId],
    queryFn: () => instagramApi.getStories(),
    enabled: !!userId,
  });
}

export function useInstagramAccountInsights(period = 'day') {
  return useQuery({
    queryKey: ['instagram-account-insights', period],
    queryFn: () => instagramApi.getAccountInsights(period),
  });
}

// Instagram limitations
export const INSTAGRAM_LIMITATIONS = {
  maxCaptionLength: 2200,
  maxHashtags: 30,
  maxCarouselItems: 10,
  supportedImageFormats: ['image/jpeg', 'image/png'],
  supportedVideoFormats: ['video/mp4'],
  maxImageSize: 8 * 1024 * 1024,
  maxVideoSize: 100 * 1024 * 1024,
  requiresBusinessAccount: true,
  publishSupported: false, // Without Content Publishing API approval
};

File: app/web/api/v1/instagram.py
"""
Instagram Graph API endpoints for read-only platform interactions.

Provides endpoints for:
- Profile retrieval
- Media feed
- Media insights (Business accounts)
- Stories

Note: Instagram Graph API requires Business/Creator account for full access.
Publishing is not supported without Content Publishing API approval.
"""

from flask import Blueprint, request
import requests
from typing import Optional

from app.auth.auth_decorators import require_auth
from app.auth.credential_manager import CredentialManager
from app.web.api.v1.responses import api_response, api_error
from app.core.logger import setup_logger

logger = setup_logger(__name__)

instagram_bp = Blueprint("instagram", __name__, url_prefix="/instagram")

# Instagram Graph API base URL
GRAPH_API_BASE = "https://graph.instagram.com"
GRAPH_API_VERSION = "v18.0"


def _get_instagram_credentials(user_id: int) -> tuple[str, str]:
    """
    Get Instagram access token and user ID for authenticated user.

    Args:
        user_id: The application user ID

    Returns:
        Tuple of (access_token, instagram_user_id)

    Raises:
        ValueError: If no credentials found or token missing
    """
    cm = CredentialManager()
    creds = cm.get_credentials(user_id, "instagram", "api")

    if not creds:
        raise ValueError("No Instagram credentials found")

    access_token = creds.get("access_token")
    instagram_user_id = creds.get("user_id")

    if not access_token:
        raise ValueError("Instagram access token not configured")

    return access_token, instagram_user_id


def _instagram_request(
    access_token: str,
    endpoint: str,
    params: Optional[dict] = None,
) -> dict:
    """
    Make authenticated request to Instagram Graph API.

    Args:
        access_token: Instagram access token
        endpoint: API endpoint path
        params: Optional query parameters

    Returns:
        JSON response from API
    """
    url = f"{GRAPH_API_BASE}/{GRAPH_API_VERSION}{endpoint}"

    request_params = params or {}
    request_params["access_token"] = access_token

    response = requests.get(url, params=request_params, timeout=30)
    response.raise_for_status()

    return response.json() if response.text else {}


def _format_profile(profile: dict) -> dict:
    """Format Instagram profile to API response."""
    return {
        "id": profile.get("id"),
        "username": profile.get("username"),
        "name": profile.get("name"),
        "biography": profile.get("biography"),
        "profile_picture_url": profile.get("profile_picture_url"),
        "website": profile.get("website"),
        "followers_count": profile.get("followers_count", 0),
        "follows_count": profile.get("follows_count", 0),
        "media_count": profile.get("media_count", 0),
        "account_type": profile.get("account_type", "PERSONAL"),
    }


def _format_media(media: dict) -> dict:
    """Format Instagram media to API response."""
    formatted = {
        "id": media.get("id"),
        "caption": media.get("caption"),
        "media_type": media.get("media_type"),
        "media_url": media.get("media_url"),
        "thumbnail_url": media.get("thumbnail_url"),
        "permalink": media.get("permalink"),
        "timestamp": media.get("timestamp"),
        "like_count": media.get("like_count"),
        "comments_count": media.get("comments_count"),
        "username": media.get("username"),
    }

    # Handle carousel albums with children
    if media.get("children") and media["children"].get("data"):
        formatted["children"] = [
            {
                "id": child.get("id"),
                "media_type": child.get("media_type"),
                "media_url": child.get("media_url"),
            }
            for child in media["children"]["data"]
        ]

    return formatted


def _format_insights(media_id: str, insights: dict) -> dict:
    """Format Instagram insights to API response."""
    result = {"media_id": media_id}

    for metric in insights.get("data", []):
        name = metric.get("name")
        values = metric.get("values", [])
        if values:
            result[name] = values[0].get("value", 0)

    return result


def _format_story(story: dict) -> dict:
    """Format Instagram story to API response."""
    return {
        "id": story.get("id"),
        "media_type": story.get("media_type"),
        "media_url": story.get("media_url"),
        "timestamp": story.get("timestamp"),
    }


@instagram_bp.route("/profile", methods=["GET"])
@require_auth
def get_own_profile():
    """
    Get the authenticated user's Instagram profile.

    Returns profile info including followers, following, media count.
    """
    try:
        access_token, user_id = _get_instagram_credentials(request.user_id)

        fields = "id,username,name,biography,profile_picture_url,website,followers_count,follows_count,media_count,account_type"
        profile = _instagram_request(
            access_token,
            f"/{user_id}",
            {"fields": fields},
        )

        return api_response(_format_profile(profile))

    except ValueError as e:
        return api_error(str(e), status=400)
    except requests.HTTPError as e:
        logger.error(f"Instagram API error fetching profile: {e}")
        error_msg = "Instagram API error"
        if e.response is not None and e.response.status_code == 404:
            error_msg = "Profile not found"
        return api_error(error_msg, status=502)
    except Exception as e:
        logger.error(f"Error fetching Instagram profile: {e}")
        return api_error("Failed to fetch profile", status=500)


@instagram_bp.route("/profile/<username>", methods=["GET"])
@require_auth
def get_profile_by_username(username: str):
    """
    Get Instagram profile by username.

    Note: Instagram Graph API doesn't support direct username lookup.
    This endpoint uses the Business Discovery API which requires
    the authenticated account to be a Business/Creator account.
    """
    try:
        access_token, user_id = _get_instagram_credentials(request.user_id)

        # Use Business Discovery API to look up other users
        fields = f"business_discovery.username({username}){{id,username,name,biography,profile_picture_url,website,followers_count,follows_count,media_count}}"
        result = _instagram_request(
            access_token,
            f"/{user_id}",
            {"fields": fields},
        )

        if "business_discovery" not in result:
            return api_error("Profile not found or not accessible", status=404)

        return api_response(_format_profile(result["business_discovery"]))

    except ValueError as e:
        return api_error(str(e), status=400)
    except requests.HTTPError as e:
        logger.error(f"Instagram API error fetching profile {username}: {e}")
        return api_error("Instagram API error", status=502)
    except Exception as e:
        logger.error(f"Error fetching Instagram profile {username}: {e}")
        return api_error("Failed to fetch profile", status=500)


@instagram_bp.route("/media", methods=["GET"])
@require_auth
def get_media():
    """
    Get the authenticated user's media feed.

    Query params:
        limit: Number of items to return (default 25, max 100)
        after: Pagination cursor
    """
    try:
        access_token, user_id = _get_instagram_credentials(request.user_id)

        limit = min(int(request.args.get("limit", 25)), 100)
        after = request.args.get("after")

        fields = "id,caption,media_type,media_url,thumbnail_url,permalink,timestamp,like_count,comments_count,children{id,media_type,media_url}"
        params = {"fields": fields, "limit": limit}
        if after:
            params["after"] = after

        result = _instagram_request(access_token, f"/{user_id}/media", params)

        media_list = [_format_media(m) for m in result.get("data", [])]

        response_data = {"data": media_list}
        if result.get("paging", {}).get("cursors", {}).get("after"):
            response_data["next_cursor"] = result["paging"]["cursors"]["after"]

        return api_response(response_data)

    except ValueError as e:
        return api_error(str(e), status=400)
    except requests.HTTPError as e:
        logger.error(f"Instagram API error fetching media: {e}")
        return api_error("Instagram API error", status=502)
    except Exception as e:
        logger.error(f"Error fetching Instagram media: {e}")
        return api_error("Failed to fetch media", status=500)


@instagram_bp.route("/media/<media_id>", methods=["GET"])
@require_auth
def get_single_media(media_id: str):
    """Get a single media item by ID."""
    try:
        access_token, _ = _get_instagram_credentials(request.user_id)

        fields = "id,caption,media_type,media_url,thumbnail_url,permalink,timestamp,like_count,comments_count,children{id,media_type,media_url}"
        result = _instagram_request(access_token, f"/{media_id}", {"fields": fields})

        return api_response(_format_media(result))

    except ValueError as e:
        return api_error(str(e), status=400)
    except requests.HTTPError as e:
        logger.error(f"Instagram API error fetching media {media_id}: {e}")
        if e.response is not None and e.response.status_code == 404:
            return api_error("Media not found", status=404)
        return api_error("Instagram API error", status=502)
    except Exception as e:
        logger.error(f"Error fetching Instagram media {media_id}: {e}")
        return api_error("Failed to fetch media", status=500)


@instagram_bp.route("/media/<media_id>/insights", methods=["GET"])
@require_auth
def get_media_insights(media_id: str):
    """
    Get insights for a media item.

    Note: Only available for Business/Creator accounts.
    Metrics vary by media type (image, video, carousel, reel).
    """
    try:
        access_token, _ = _get_instagram_credentials(request.user_id)

        # Different metrics for different media types
        # Default to image metrics, caller can specify via query param
        media_type = request.args.get("media_type", "IMAGE").upper()

        if media_type == "VIDEO" or media_type == "REELS":
            metric = "impressions,reach,video_views,saved"
        elif media_type == "CAROUSEL_ALBUM":
            metric = "impressions,reach,carousel_album_engagement,carousel_album_reach"
        else:
            metric = "impressions,reach,engagement,saved"

        result = _instagram_request(
            access_token,
            f"/{media_id}/insights",
            {"metric": metric},
        )

        return api_response(_format_insights(media_id, result))

    except ValueError as e:
        return api_error(str(e), status=400)
    except requests.HTTPError as e:
        logger.error(f"Instagram API error fetching insights for {media_id}: {e}")
        if e.response is not None:
            if e.response.status_code == 400:
                return api_error("Insights not available for this media or account type", status=400)
            if e.response.status_code == 404:
                return api_error("Media not found", status=404)
        return api_error("Instagram API error", status=502)
    except Exception as e:
        logger.error(f"Error fetching Instagram insights for {media_id}: {e}")
        return api_error("Failed to fetch insights", status=500)


@instagram_bp.route("/stories", methods=["GET"])
@require_auth
def get_stories():
    """
    Get the authenticated user's active stories.

    Note: Only returns stories that are still active (within 24 hours).
    Requires Business/Creator account.
    """
    try:
        access_token, user_id = _get_instagram_credentials(request.user_id)

        fields = "id,media_type,media_url,timestamp"
        result = _instagram_request(access_token, f"/{user_id}/stories", {"fields": fields})

        stories = [_format_story(s) for s in result.get("data", [])]

        return api_response({"data": stories})

    except ValueError as e:
        return api_error(str(e), status=400)
    except requests.HTTPError as e:
        logger.error(f"Instagram API error fetching stories: {e}")
        return api_error("Instagram API error", status=502)
    except Exception as e:
        logger.error(f"Error fetching Instagram stories: {e}")
        return api_error("Failed to fetch stories", status=500)


@instagram_bp.route("/account/insights", methods=["GET"])
@require_auth
def get_account_insights():
    """
    Get account-level insights.

    Query params:
        period: Time period (day, week, days_28, lifetime)
        metric: Specific metrics to retrieve

    Note: Only available for Business/Creator accounts.
    """
    try:
        access_token, user_id = _get_instagram_credentials(request.user_id)

        period = request.args.get("period", "day")
        metric = request.args.get(
            "metric",
            "impressions,reach,follower_count,profile_views",
        )

        result = _instagram_request(
            access_token,
            f"/{user_id}/insights",
            {"metric": metric, "period": period},
        )

        # Format insights data
        insights = {}
        for item in result.get("data", []):
            name = item.get("name")
            values = item.get("values", [])
            if values:
                insights[name] = values[0].get("value", 0)

        return api_response({"period": period, "insights": insights})

    except ValueError as e:
        return api_error(str(e), status=400)
    except requests.HTTPError as e:
        logger.error(f"Instagram API error fetching account insights: {e}")
        if e.response is not None and e.response.status_code == 400:
            return api_error("Account insights not available (requires Business account)", status=400)
        return api_error("Instagram API error", status=502)
    except Exception as e:
        logger.error(f"Error fetching Instagram account insights: {e}")
        return api_error("Failed to fetch account insights", status=500)
